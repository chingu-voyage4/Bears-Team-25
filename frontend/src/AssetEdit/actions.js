const save = payload =>
  fetch('/api/v1/asset', {
    method: 'post',
    credentials: 'same-origin',
    body: payload,
  }).then(res => res.json);

export default { save };
