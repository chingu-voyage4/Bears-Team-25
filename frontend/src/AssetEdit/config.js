const fileTypes = ['audio', 'image', 'video', 'other'];
const maxFileSizeMb = 20;

export { fileTypes, maxFileSizeMb };
