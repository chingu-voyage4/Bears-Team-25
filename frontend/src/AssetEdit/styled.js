import styled from 'styled-components';

export const Button = styled.button`
  background-color: #15df89;
  border: none;
  border-radius: 12px;
  box-shadow: 0 1px 2px #aaa;
  color: #fff;
  cursor: pointer;
  display: block;
  font-size: 1.2rem;
  font-weight: bold;
  height: auto;
  margin: 1rem;
  max-width: 10rem;
  padding: 0.5rem 1.2rem;
  position: relative;

  &:hover {
    background-color: #08d27c;
  }

  &:active {
    box-shadow: none;
    top: 1px;
  }
`;

Button.displayName = 'Button';

export const DropArea = styled.button`
  background-color: #e2e2e2;
  border: none;
  box-shadow: 0 0 2px #333;
  height: 30vh;
  margin: auto;
  padding: 0;
  padding-top: 1rem;
  position: relative;
  width: 40%;
`;

DropArea.displayName = 'DropArea';

export const Hint = styled.p`
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
`;

export const ImgPreview = styled.img`
  height: 75%;
  transform: ${p => (p.scaled ? 'scale(2) translateY(25%)' : 'scale(1)')};
  transition: transform 0.15s ease;
`;

ImgPreview.displayName = 'ImgPreview';

export const InvisibleInput = styled.input.attrs({
  type: 'file',
})`
  display: none;
`;

InvisibleInput.displayName = 'InvisibleInput';

export const InputField = styled.p`
  align-items: center;
  background-color: #e2e2e2;
  border: 1px solid gray;
  display: flex;
  flex: 1;
  font-size: 1.1rem;
  font-weight: normal;
  height: 100%;
  line-height: 1.1rem;
  margin-left: 2rem;
  margin-right: 30%;
  max-width: 50%;
  padding-left: 0.5rem;
`;

InputField.displayName = 'InputField';

export const Label = styled.label`
  align-items: center;
  display: flex;
  font-weight: bold;
  height: 2rem;
  justify-content: flex-end;
  margin-bottom: 2rem;
`;

Label.displayName = 'Label';

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  margin: auto;
  max-width: 75rem;
  padding: 2rem 0;
`;
