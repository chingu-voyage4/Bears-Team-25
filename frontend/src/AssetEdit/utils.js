const getFileType = (file, availableTypes = []) => {
  if (
    !file ||
    !file.type ||
    (typeof availableTypes !== 'object' && !availableTypes.length)
  ) {
    return 'other';
  }

  let fileType = file.type.split('/')[0] || 'other';

  if (!availableTypes.includes(fileType)) {
    fileType = 'other';
  }

  return fileType;
};

const getMediaType = d64 => {
  if (!d64) {
    return '';
  }

  if (typeof d64 !== 'string') return 'other';

  const match = d64.match(/(:)(\w+)(\/)/);

  if (!match || match.length < 3) {
    return 'other';
  }

  return match[2];
};

const readFile = file =>
  new Promise(resolve => {
    if (!(file instanceof File)) {
      resolve('');
    }

    const reader = new FileReader();

    reader.addEventListener('load', () => {
      resolve(reader.result);
    });

    reader.readAsDataURL(file);
  });

const getLocalUrl = blob => URL.createObjectURL(blob);

const packageData = ({ _id, file, title, description }) => {
  const payload = new FormData();

  if (_id) {
    payload.append('_id', _id);
  }

  payload.append('title', title);
  payload.append('description', description);

  if (file) {
    payload.append('blob', file);
  }

  return payload;
};

export { getFileType, getMediaType, packageData, readFile, getLocalUrl };
