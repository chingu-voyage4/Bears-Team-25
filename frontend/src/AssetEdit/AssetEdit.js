/* eslint-disable react/require-default-props */
import React from 'react';
import PropTypes from 'prop-types';
import actions from './actions';

import Input from './Input';
import {
  Button,
  DropArea,
  Hint,
  ImgPreview,
  InputField,
  InvisibleInput,
  Label,
  Wrapper,
} from './styled';

import { fileTypes, maxFileSizeMb } from './config';
import { getFileType, readFile } from './utils';

const MAX_FILE_SIZE_MB = maxFileSizeMb;

class AssetEdit extends React.Component {
  static propTypes = {
    id: PropTypes.string,
  };
  state = {
    fileType: null,
    file: null,
    data64: null,
    scale: false,
    description: '',
    title: '',
    localUrl: null,
    _id: null,
    creator: null,
  };
  // eslint-disable-next-line react/sort-comp
  getAsset = id => {
    const promises = [];
    promises.push(
      actions.get(id).then(json => {
        const { title, description, content_type: type, creator } = json;
        const fileType = getFileType({ type }, fileTypes);
        return { title, description, fileType, creator };
      })
    );
    promises.push(actions.getContent(this.props.id).then(localUrl => localUrl));
    return promises;
  };
  componentDidMount = () => {
    if (this.props.id) {
      Promise.all(this.getAsset(this.props.id)).then(results => {
        const [detail, localUrl] = results;
        this.setState({
          _id: this.props.id,
          scale: false,
          ...detail,
          localUrl,
          file: null,
          data64: null,
        });
      });
    }
  };

  handleSelect = e => {
    const fileType = e.target.value;
    this.setState(() => ({
      fileType,
    }));
  };

  handleChange = e => {
    if (!e.target.files.length) {
      return;
    }

    const file = e.target.files[0];

    if (file.size >= MAX_FILE_SIZE_MB * 1e6) {
      // eslint-disable-next-line
      return alert(`File too big. Max.file size: ${MAX_FILE_SIZE_MB}MB`);
    }

    this.setState(() => ({
      fileType: getFileType(file, fileTypes),
      file,
    }));
    this.set64(file);
  };

  set64 = file => {
    readFile(file).then(data64 =>
      this.setState(() => ({ data64, localUrl: null }))
    );
  };

  handleDrop = e => {
    e.preventDefault();
    let file;

    if (e.dataTransfer.items) {
      if (e.dataTransfer.items[0].kind === 'file') {
        file = e.dataTransfer.items[0].getAsFile();
        e.dataTransfer.items.clear();
      }
    } else {
      [file] = e.dataTransfer.files;
      e.dataTransfer.clearData();
    }

    this.setState(() => ({
      fileType: getFileType(file, fileTypes),
      file,
    }));
    this.set64(file);
  };

  handleDrag = e => {
    e.preventDefault();
  };

  // eslint-disable-next-line
  handleDropAreaClick = e => {
    if (e.target.tagName !== 'IMG') {
      return this.fileInput.click();
    }

    this.setState(s => ({
      scale: !s.scale,
    }));
  };
  handleFieldChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };
  packageData = () => {
    const { file, _id, title, description } = this.state;
    const payload = new FormData();
    if (_id) {
      payload.append('_id', _id);
    }
    payload.append('title', title);
    payload.append('description', description);
    if (file) {
      payload.append('blob', file);
    }
    return payload;
  };
  handleSave = () => {
    const payload = this.packageData();
    actions.save(payload).then(json => {
      this.setState({ _id: json._id });
    });
  };

  renderAsset = (type, src) => {
    switch (true) {
      case type === 'image':
        return <ImgPreview src={src} alt="Preview" scaled={this.state.scale} />;
      case type === 'audio':
        // eslint-disable-next-line
        return <audio src={src} controls />;
      case type === 'video':
        // eslint-disable-next-line
        return <video style={{ height: '75%' }} src={src} controls />;
      default:
        return <div>No preview available</div>;
    }
  };
  render() {
    const {
      description,
      title,
      localUrl,
      fileType,
      data64,
      creator,
    } = this.state;
    let username = '';
    if (creator) {
      // eslint-disable-next-line prefer-destructuring
      username = creator.username;
    }
    const embedUrl = `//api/v1/asset/content/${this.state._id}`;
    return (
      <Wrapper>
        <Label htmlFor="creator">
          Owner:
          <InputField>{username}</InputField>
        </Label>
        <Input
          value={title}
          label="Title:"
          name="title"
          onChange={this.handleFieldChange}
        />
        <Input
          value={description}
          label="Description:"
          name="description"
          onChange={this.handleFieldChange}
        />
        <Label htmlFor="asset">
          Asset type:
          <InputField>{fileType}</InputField>
        </Label>
        <InvisibleInput
          innerRef={ref => {
            this.fileInput = ref;
          }}
          onChange={this.handleChange}
        />
        <Label htmlFor="embed url">
          Embed Url:
          <InputField>{embedUrl}</InputField>
        </Label>
        <DropArea
          onClick={this.handleDropAreaClick}
          onDrop={this.handleDrop}
          onDragOver={this.handleDrag}
        >
          <Hint>Drop file or click area to select from disk</Hint>
          {data64 && this.renderAsset(fileType, data64)}
          {localUrl && this.renderAsset(fileType, localUrl)}
        </DropArea>
        <Button onClick={this.handleSave}>Save</Button>
      </Wrapper>
    );
  }
}

export default AssetEdit;
