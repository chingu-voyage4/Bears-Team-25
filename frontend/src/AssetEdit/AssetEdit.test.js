import React from 'react';
import { mount, shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import AssetEdit from '.';

describe('AssetEdit component', () => {
  it('matches snapshot', () => {
    const tree = renderer.create(<AssetEdit />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders image preview', () => {
    const props = {
      asset: 'data:image/jpeg;base64,bG9yZW0gaXBzdW0=',
      description: 'descr.',
      fileType: 'image',
      title: 'title',
    };

    const wrapper = shallow(<AssetEdit {...props} />);

    const dropArea = wrapper.find('DropArea');

    expect(dropArea.find('ImgPreview').length).toEqual(1);
    expect(dropArea.find('audio').length).toEqual(0);
    expect(dropArea.find('video').length).toEqual(0);
  });

  it('renders audio preview', () => {
    const props = {
      asset: 'data:audio/mp3;base64,bG9yZW0gaXBzdW0=',
      description: 'descr.',
      fileType: 'audio',
      title: 'title',
    };

    const wrapper = shallow(<AssetEdit {...props} />);

    const dropArea = wrapper.find('DropArea');

    expect(dropArea.find('ImgPreview').length).toEqual(0);
    expect(dropArea.find('audio').length).toEqual(1);
    expect(dropArea.find('video').length).toEqual(0);
  });

  it('renders video preview', () => {
    const props = {
      asset: 'data:video/mp4;base64,bG9yZW0gaXBzdW0=',
      description: 'descr.',
      fileType: 'audio',
      title: 'title',
    };

    const wrapper = shallow(<AssetEdit {...props} />);

    const dropArea = wrapper.find('DropArea');

    expect(dropArea.find('ImgPreview').length).toEqual(0);
    expect(dropArea.find('audio').length).toEqual(0);
    expect(dropArea.find('video').length).toEqual(1);
  });

  it('handles DropArea click', () => {
    const mockFn = jest.fn();
    const wrapper = mount(<AssetEdit />);

    wrapper.instance().fileInput.click = mockFn;
    const initialScale = wrapper.state('scale');
    wrapper.find('DropArea').simulate('click');
    const newScale = wrapper.state('scale');

    expect(mockFn).toHaveBeenCalled();
    expect(initialScale).toEqual(newScale);
  });

  it('handles DropArea click with image preview', () => {
    const props = {
      asset: 'data:image/jpeg;base64,bG9yZW0gaXBzdW0=',
      description: 'descr.',
      fileType: 'image',
      title: 'title',
    };

    const mockFn = jest.fn();
    const wrapper = mount(<AssetEdit {...props} />);
    const initialScale = wrapper.state('scale');

    wrapper.instance().fileInput.click = mockFn;
    wrapper.find('ImgPreview').simulate('click');
    const newScale = wrapper.state('scale');

    expect(mockFn).not.toHaveBeenCalled();
    expect(newScale).toEqual(!initialScale);
  });
});
