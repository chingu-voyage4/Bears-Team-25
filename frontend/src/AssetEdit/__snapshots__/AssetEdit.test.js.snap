// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetEdit component matches snapshot 1`] = `
.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1.1rem;
  height: 100%;
  margin-left: 2rem;
  margin-right: 30%;
  max-width: 50%;
  padding-left: 0.5rem;
}

.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: bold;
  height: 2rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-bottom: 2rem;
}

.c8 {
  background-color: #15df89;
  border: none;
  border-radius: 12px;
  box-shadow: 0 1px 2px #aaa;
  color: #fff;
  cursor: pointer;
  display: block;
  font-size: 1.2rem;
  font-weight: bold;
  height: auto;
  margin: 1rem;
  max-width: 10rem;
  padding: 0.5rem 1.2rem;
  position: relative;
}

.c8:hover {
  background-color: #08d27c;
}

.c8:active {
  box-shadow: none;
  top: 1px;
}

.c6 {
  background-color: #e2e2e2;
  border: none;
  box-shadow: 0 0 2px #333;
  height: 30vh;
  margin: auto;
  padding: 0;
  padding-top: 1rem;
  position: relative;
  width: 40%;
}

.c7 {
  left: 0;
  position: absolute;
  right: 0;
  top: 0;
}

.c5 {
  display: none;
}

.c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #e2e2e2;
  border: 1px solid gray;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 1.1rem;
  font-weight: normal;
  height: 100%;
  line-height: 1.1rem;
  margin-left: 2rem;
  margin-right: 30%;
  max-width: 50%;
  padding-left: 0.5rem;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: bold;
  height: 2rem;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-bottom: 2rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: auto;
  max-width: 75rem;
  padding: 2rem 0;
}

<div
  className="c0"
>
  <label
    className="c1"
    htmlFor="title"
  >
    Title:
    <input
      className="c2"
      name="title"
      onChange={[Function]}
      value=""
    />
  </label>
  <label
    className="c1"
    htmlFor="description"
  >
    Description:
    <input
      className="c2"
      name="description"
      onChange={[Function]}
      value=""
    />
  </label>
  <label
    className="c3"
    htmlFor="asset"
  >
    Asset type:
    <p
      className="c4"
    />
  </label>
  <input
    className="c5"
    onChange={[Function]}
    type="file"
  />
  <label
    className="c3"
    htmlFor="embed url"
  >
    Embed Url:
    <p
      className="c4"
    >
      //api/v1/asset/content/null
    </p>
  </label>
  <button
    className="c6"
    onClick={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <p
      className="c7"
    >
      Drop file or click area to select from disk
    </p>
  </button>
  <button
    className="c8"
    onClick={[Function]}
  >
    Save
  </button>
</div>
`;
